File: easyandroidform/src/test/java/com/sa/easyandroidform/field_view/InputFieldView.java
package com.sa.easyandroidform.field_view;

import android.content.Context;
import android.widget.EditText;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class InputFieldView extends BaseInputFieldView<Integer> {

    public InputFieldView(@NotNull Context context) {
        super(context);
    }

    @NotNull
    @Override
    public EditText getEditText() {
        return null;
    }

    @Override
    public Integer resolveFrom(@NotNull CharSequence charSequence) {
        return Integer.valueOf(charSequence.toString());
    }

    @Override
    public CharSequence resolve(@Nullable Integer value) {
        if (value != null)
            return String.valueOf(value);
        return null;
    }

    @Override
    public boolean isSame(Integer value, Integer prevValue) {
        return value.equals(prevValue);
    }

    @Override
    public void fieldMandatory() { }

    @Override
    public void displayError(boolean show, @Nullable String error) { }
}

File: app/src/main/java/com/sa/easyandroidfrom/example/InputFieldView.kt
package com.sa.easyandroidfrom.example

import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import android.view.View
import android.widget.EditText
import com.sa.easyandroidfrom.ObjectUtils
import com.sa.easyandroidfrom.field_view.BaseInputFieldView
import kotlinx.android.synthetic.main.input_field_view.view.*


abstract class InputFieldView<F> @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0) :
    BaseInputFieldView<F>(context, attrs, defStyleAttr) {

    init {
        View.inflate(context, R.layout.input_field_view, this)
        val handler = Handler()
        handler.postDelayed( {
            editTextLayout.isEndIconVisible = false
        }, 1000)
        editText.setOnFocusChangeListener { v, hasFocus ->
            editTextLayout.isEndIconVisible = hasFocus
        }
    }


    final override fun getEditText(): EditText {
        return editText
    }

    override fun fieldMandatory() {
    }

    override fun error() {
    }

    override fun edited() {
    }

    override fun neutral() {
    }

    override fun displayError(show: Boolean, error: String?) {
        editTextLayout.error = error
        editTextLayout.isErrorEnabled = show
    }

    final override fun isSame(value: F?, prevValue: F?): Boolean {

        if (ObjectUtils.isNull(value) && ObjectUtils.isNull(prevValue)) {
            return true
        } else if (ObjectUtils.isNotNull(value) && ObjectUtils.isNull(prevValue)) {
            return false
        } else if(value != null && prevValue != null){
            return compare(value, prevValue)
        }
        return false
    }

    abstract fun compare(value: F, prevValue: F): Boolean

}
